File: examples/python/named_actors.py
#!/usr/bin/env python3
"""
Named Actors Example

Named actors can be discovered by name instead of specific ActorId,
enabling service discovery.

Usage: python examples/python/named_actors.py
"""

import asyncio

import pulsing as pul


@pul.remote
class EchoActor:
    """Simple echo actor that can be discovered by name."""

    def echo(self, message: str) -> dict[str, str]:
        print(f"[Echo] {message}")
        return {"echo": message}


async def main():
    print("=== Pulsing Named Actors ===\n")

    await pul.init()
    try:
        print("✓ System started\n")

        await EchoActor.spawn(name="echo")
        print("✓ Created: echo (named, discoverable)\n")

        print("--- Resolve by name ---")
        actor = await EchoActor.resolve("echo")
        resp = await actor.echo("Hello!")
        print(f"Response: {resp['echo']}\n")

        print("✓ Done!")
    finally:
        await pul.shutdown()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/rust/named_actors.rs
//! Named Actors Example - Service Discovery
//!
//! Named actors can be discovered by path (e.g., "services/echo")
//! instead of ActorId, enabling service discovery across cluster.
//!
//! Run: cargo run --example named_actors -p pulsing-actor

use pulsing_actor::prelude::*;

struct Echo;

#[async_trait]
impl Actor for Echo {
    async fn receive(&mut self, msg: Message, _ctx: &mut ActorContext) -> anyhow::Result<Message> {
        let s: String = msg.unpack()?;
        Message::pack(&format!("echo: {}", s))
    }
}

#[tokio::main]
async fn main() -> anyhow::Result<()> {
    let system = ActorSystem::builder().build().await?;

    // Spawn named actor - name is now the full path
    system.spawn_named("services/echo", Echo).await?;

    // Resolve by name and send message
    let actor = system.resolve("services/echo").await?;
    let resp: String = actor.ask("hello".to_string()).await?;
    println!("{}", resp);

    system.shutdown().await
}
